### PR TITLE
break nptables on block-level structures

### PR DESCRIPTION
**Pull request by UziTech**
**Marked version:** master

**Markdown flavor:** GitHub Flavored Markdown

## Description

This PR copies #1598 to add breaking block-level rules to `nptable` (tables that don't start or end with a pipe)

e.g:
```markdown
# nptable

 abc | def
 --- | ---
  1  |  2
  3  |  4

# table

| abc | def |
| --- | --- |
|  1  |  2  |
|  3  |  4  |
```

I also removed allowing setext header from breaking a table since it doesn't break on github. We can change it later easily if github changes. see https://github.com/markedjs/marked/pull/1598#pullrequestreview-355648462

## Contributor

- [x] Test(s) exist to ensure functionality and minimize regression (if no tests added, list tests covering this PR).

## Committer

In most cases, this should be a different person than the contributor.

- [ ] Draft GitHub release notes have been updated.
- [ ] CI is green (no forced merge required).
- [ ] Merge PR

### CODE REVIEW

File: src/rules.js [hunk 1]
```diff
@@ -93,22 +93,38 @@ block.normal = merge({}, block);
...
+  .replace('list', ' {0,3}(?:[*+-]|1[.)]) ') // only lists starting from 1 can interrupt
+  .replace('html', '</?(?:tag)(?: +|\\n|/?>)|<(?:script|pre|style|!--)')
+  .replace('tag', block._tag) // tables can be interrupted by type (6) html blocks
+  .getRegex();
+
 block.gfm.table = edit(block.gfm.table)
   .replace('hr', block.hr)
   .replace('heading', ' {0,3}#{1,6} ')
-  .replace('lheading', '([^\\n]+)\\n {0,3}(=+|-+) *(?:\\n+|$)')
+  // .replace('lheading', '([^\\n]+)\\n {0,3}(=+|-+) *(?:\\n+|$)')
```
Comment (styfle) [inline src/rules.js]: I think we should use git log history instead of leaving a comment like this

File: src/rules.js [hunk 2]
```diff
@@ -93,22 +93,38 @@ block.normal = merge({}, block);
...
   table: '^ *\\|(.+)\\n' // Header
     + ' *\\|?( *[-:]+[-| :]*)' // Align
-    + '(?:\\n((?:(?!^|>|\\n| |hr|heading|lheading|code|fences|list|html).*(?:\\n|$))*)\\n*|$)' // Cells
+    + '(?:\\n *((?:(?!\\n|hr|heading|lheading|blockquote|code|fences|list|html).*(?:\\n|$))*)\\n*|$)' // Cells
 });
 
+block.gfm.nptable = edit(block.gfm.nptable)
+  .replace('hr', block.hr)
+  .replace('heading', ' {0,3}#{1,6} ')
+  // .replace('lheading', '([^\\n]+)\\n {0,3}(=+|-+) *(?:\\n+|$)')
```
Comment (styfle) [inline src/rules.js]: I think we should use git log history instead of leaving a comment like this

**Comment by UziTech** [comment]
@joshbruce could you review this so we can release [v0.8.1](https://github.com/markedjs/marked/pull/1616)

**Comment by joshbruce** [comment]
So glad to see all the tests.

**Pull request merged**